File: java/03-bootstrap/src/test/java/com/fResult/bootstrap/TransactionTestMixin.java
package com.fResult.bootstrap;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

import com.fResult.bootstrap.customers.services.CustomerService;
import org.slf4j.LoggerFactory;

public interface TransactionTestMixin {
  default void testTransactionalityOfSave(CustomerService customerService) {
    final var log = LoggerFactory.getLogger(getClass());
    log.info("using customer {}", customerService.getClass().getSimpleName());

    final var expectedResult = customerService.findAll();
    final var expectedCount = expectedResult.size();

    try {
      customerService.save("Anderson", null);
    } catch (Exception ex) {
      final var actualResult = customerService.findAll();
      assertEquals(
          expectedCount,
          actualResult.size(),
          "there should be no new records in the database: " + ex.getMessage());
      return;
    }

    fail("Expected exception was not thrown");
  }
}

File: kotlin/03-bootstrap/src/test/kotlin/com/fResult/bootstrap/TransactionTestMixin.kt
package com.fResult.bootstrap

import com.fResult.bootstrap.customers.services.CustomerService
import org.apache.logging.log4j.LogManager
import org.junit.jupiter.api.Assertions.assertEquals

interface TransactionTestMixin {
  fun testTransactionalityOfSave(customerService: CustomerService) {
    val log = LogManager.getLogger(javaClass)
    log.info("using customer {}", customerService.javaClass.simpleName)
    val count = customerService.findAll().size

    try {
      customerService.save("Bob", "null")
    } catch (ex: Exception) {
      assertEquals(
        count, customerService.findAll().size,
        "there should be no new records in the database",
      )
      return
    }

    // fail<Exception>("Expected exception was not thrown")
  }
}
